### PR TITLE
cli: add shell completion

### PR DESCRIPTION
**Pull request by crazy-max**
closes #1674 
partially solves #1072

Add completion to:
* List builder names for `--builder` persistent flag and as arg for some cmds
* Folders (context) for build command
* List bake targets

```
$ docker __completeNoDesc buildx --builder ""
builder
builder2
buildkit0106
buildkit093
remote-builder
default
desktop-linux
docker201016
docker2300
docker2400
:4
Completion ended with directive: ShellCompDirectiveNoFileComp
```

```
$ docker __completeNoDesc buildx --builder "do"
docker201016
docker2300
docker2400
:4
Completion ended with directive: ShellCompDirectiveNoFileComp
```

For bake, can be tested within this repo:

```
$ docker __completeNoDesc buildx ""
bake
build
create
du
help
imagetools
inspect
ls
prune
rm
stop
use
version
:4
Completion ended with directive: ShellCompDirectiveNoFileComp
```

```
$ docker __completeNoDesc buildx bake ""
default
validate
meta-helper
_common
lint
validate-vendor
validate-docs
validate-authors
validate-generated-files
update-vendor
update-docs
update-authors
update-generated-files
mod-outdated
test
binaries
binaries-cross
release
image
image-cross
image-local
:4
Completion ended with directive: ShellCompDirectiveNoFileComp
```

```
$ docker __completeNoDesc buildx bake "v"
validate
validate-vendor
validate-docs
validate-authors
validate-generated-files
:4
Completion ended with directive: ShellCompDirectiveNoFileComp
```

### CODE REVIEW

File: util/cobrautil/completion.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+// ValidArgsFn defines a completion func to be returned to fetch completion options
+type ValidArgsFn func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective)
+
+func NoCompletion(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return nil, cobra.ShellCompDirectiveNoSpace
+}
+
+func CompleteBakeTargets(files []string) ValidArgsFn {
```
Comment (thaJeztah) [inline util/cobrautil/completion.go:20]: Perhaps bit if a nit, but if these were in a package named "completion", then these could be `completion.BakeTargets` (etc) which may a bit shorter, and more descriptive

File: util/cobrautil/completion.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			return nil, cobra.ShellCompDirectiveError
+		}
+		var filtered []string
+		if toComplete == "" {
+			return tgts, cobra.ShellCompDirectiveNoFileComp
+		}
+		for _, tgt := range tgts {
+			if strings.HasPrefix(tgt, toComplete) {
+				filtered = append(filtered, tgt)
+			}
```
Comment (jedevc) [inline util/cobrautil/completion.go:37]: Could we split out the filtering logic into a separate function to share with other completion funcs as well?

Comment (crazy-max) [reply util/cobrautil/completion.go]: Don't think it's worth it currently. Also targets and builders are not of the same type so would need some reflection.

Comment (thaJeztah) [reply util/cobrautil/completion.go]: Generics! 😂

File: commands/build.go [hunk 3]
```diff
@@ -253,6 +253,9 @@ func buildCmd(dockerCli command.Cli, rootOpts *rootOptions) *cobra.Command {
 			}
 			return runBuild(dockerCli, options)
 		},
+		ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+			return nil, cobra.ShellCompDirectiveFilterDirs
+		},
```
Comment (jedevc) [inline commands/build.go:258]: Looks like this one was missed from the `completion` package?

Comment (crazy-max) [reply commands/build.go:258]: As it just returns `cobra.ShellCompDirectiveFilterDirs` I don't think we need a specific func in completion pkg if this is what you mean?

**Review by jedevc — APPROVED** [review]

**Pull request merged**